File: apps/android/app/src/main/java/com/kris99/baozi/android/ui/home/OnboardingCoachmarks.kt
package com.kris99.baozi.android.ui.home

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.Shadow
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.kris99.baozi.android.ui.BaoziTheme
import kotlin.math.PI
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sin
import kotlin.math.sqrt

internal enum class CoachmarkTarget {
    AddServer,
    NewThread,
    Search,
    Voice,
}

private enum class CoachmarkLineStyle {
    SmoothCurve,
    SolidSquiggle,
    DashedSquiggle,
    Dotted,
}

private enum class CoachmarkLabelAlignment {
    Leading,
    Center,
    Trailing,
}

private data class CoachmarkItem(
    val id: CoachmarkTarget,
    val primary: String,
    val secondary: String?,
    val positionX: Float,
    val positionY: Float,
    val labelWidth: Dp,
    val labelAlignment: CoachmarkLabelAlignment,
    val style: CoachmarkLineStyle,
    val isPrimary: Boolean,
)

private val coachmarkItems = listOf(
    CoachmarkItem(
        id = CoachmarkTarget.AddServer,
        primary = "添加远程电脑",
        secondary = "有的话再加",
        positionX = 0.55f,
        positionY = 0.20f,
        labelWidth = 200.dp,
        labelAlignment = CoachmarkLabelAlignment.Center,
        style = CoachmarkLineStyle.SmoothCurve,
        isPrimary = false,
    ),
    CoachmarkItem(
        id = CoachmarkTarget.Search,
        primary = "查看\n全部会话",
        secondary = null,
        positionX = 0.92f,
        positionY = 0.62f,
        labelWidth = 110.dp,
        labelAlignment = CoachmarkLabelAlignment.Trailing,
        style = CoachmarkLineStyle.DashedSquiggle,
        isPrimary = false,
    ),
    CoachmarkItem(
        id = CoachmarkTarget.NewThread,
        primary = "新建会话",
        secondary = "或直接输入消息",
        positionX = 0.50f,
        positionY = 0.70f,
        labelWidth = 220.dp,
        labelAlignment = CoachmarkLabelAlignment.Center,
        style = CoachmarkLineStyle.SolidSquiggle,
        isPrimary = true,
    ),
    CoachmarkItem(
        id = CoachmarkTarget.Voice,
        primary = "实时语音",
        secondary = "需在设置填 OpenAI 密钥",
        positionX = 0.20f,
        positionY = 0.78f,
        labelWidth = 160.dp,
        labelAlignment = CoachmarkLabelAlignment.Leading,
        style = CoachmarkLineStyle.Dotted,
        isPrimary = false,
    ),
)

@Composable
internal fun OnboardingCoachmarks(
    targets: Map<CoachmarkTarget, Rect>,
    modifier: Modifier = Modifier,
) {
    BoxWithConstraints(modifier = modifier) {
        val transition = rememberInfiniteTransition(label = "coachmarkHalo")
        val haloPhase by transition.animateFloat(
            initialValue = 0f,
            targetValue = 1f,
            animationSpec = infiniteRepeatable(
                animation = tween(durationMillis = 1600),
                repeatMode = RepeatMode.Reverse,
            ),
            label = "coachmarkHaloPhase",
        )
        val density = LocalDensity.current
        val widthPx = with(density) { maxWidth.toPx() }
        val heightPx = with(density) { maxHeight.toPx() }
        val labelHeightPx = with(density) { 60.dp.toPx() }
        val clampPx = with(density) { 8.dp.toPx() }
        val voiceFallback = rememberVoiceFallback(widthPx, heightPx, density)
        val resolved = coachmarkItems.mapNotNull { item ->
            val target = targets[item.id] ?: if (item.id == CoachmarkTarget.Voice) voiceFallback else null
            target?.let { item to it }
        }
        val labelRects = resolved.associate { (item, _) ->
            val labelWidthPx = with(density) { item.labelWidth.toPx() }
            val proposedX = widthPx * item.positionX - labelWidthPx / 2f
            val proposedY = heightPx * item.positionY - labelHeightPx / 2f
            val clampedX = proposedX.coerceIn(clampPx, widthPx - labelWidthPx - clampPx)
            val clampedY = proposedY.coerceIn(clampPx, heightPx - labelHeightPx - clampPx)
            item.id to Rect(clampedX, clampedY, clampedX + labelWidthPx, clampedY + labelHeightPx)
        }

        Canvas(modifier = Modifier.matchParentSize()) {
            resolved.forEach { (item, targetRect) ->
                val labelRect = labelRects[item.id] ?: return@forEach
                if (item.isPrimary) {
                    drawCoachmarkHalo(targetRect, haloPhase)
                }
                drawCoachmarkArrow(
                    from = labelRect.center,
                    to = targetRect.center,
                    targetRect = targetRect,
                    labelRect = labelRect,
                    style = item.style,
                    isPrimary = item.isPrimary,
                )
            }
        }

        resolved.forEach { (item, _) ->
            val labelRect = labelRects[item.id] ?: return@forEach
            CoachmarkLabel(
                item = item,
                modifier = Modifier
                    .offset {
                        IntOffset(labelRect.left.roundToInt(), labelRect.top.roundToInt())
                    }
                    .size(width = item.labelWidth, height = 60.dp),
            )
        }
    }
}

@Composable
private fun rememberVoiceFallback(
    widthPx: Float,
    heightPx: Float,
    density: androidx.compose.ui.unit.Density,
): Rect {
    return with(density) {
        val size = 44.dp.toPx()
        val leading = 14.dp.toPx()
        val bottomInset = 4.dp.toPx()
        Rect(
            left = leading,
            top = heightPx - bottomInset - size,
            right = leading + size,
            bottom = heightPx - bottomInset,
        )
    }
}

@Composable
private fun CoachmarkLabel(
    item: CoachmarkItem,
    modifier: Modifier = Modifier,
) {
    val alignment = when (item.labelAlignment) {
        CoachmarkLabelAlignment.Leading -> Alignment.TopStart
        CoachmarkLabelAlignment.Center -> Alignment.TopCenter
        CoachmarkLabelAlignment.Trailing -> Alignment.TopEnd
    }
    val textAlign = when (item.labelAlignment) {
        CoachmarkLabelAlignment.Leading -> TextAlign.Start
        CoachmarkLabelAlignment.Center -> TextAlign.Center
        CoachmarkLabelAlignment.Trailing -> TextAlign.End
    }
    Box(modifier = modifier, contentAlignment = alignment) {
        androidx.compose.foundation.layout.Column(
            horizontalAlignment = when (item.labelAlignment) {
                CoachmarkLabelAlignment.Leading -> Alignment.Start
                CoachmarkLabelAlignment.Center -> Alignment.CenterHorizontally
                CoachmarkLabelAlignment.Trailing -> Alignment.End
            },
        ) {
            val shadow = if (BaoziTheme.isDark) {
                Shadow(color = Color.Black.copy(alpha = 0.7f), offset = Offset(0f, 1f), blurRadius = 4f)
            } else {
                null
            }
            Text(
                text = item.primary,
                color = BaoziTheme.accent,
                fontSize = 12.sp,
                fontWeight = FontWeight.SemiBold,
                fontFamily = BaoziTheme.monoFont,
                textAlign = textAlign,
                style = androidx.compose.ui.text.TextStyle(shadow = shadow),
            )
            if (item.secondary != null) {
                Text(
                    text = item.secondary,
                    color = BaoziTheme.textSecondary,
                    fontSize = 10.sp,
                    fontWeight = FontWeight.Normal,
                    fontFamily = BaoziTheme.monoFont,
                    textAlign = textAlign,
                    style = androidx.compose.ui.text.TextStyle(shadow = shadow),
                )
            }
        }
    }
}

private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawCoachmarkHalo(
    targetRect: Rect,
    phase: Float,
) {
    val radius = (max(targetRect.width, targetRect.height) + 18.dp.toPx()) / 2f
    val scale = 1f + 0.08f * phase
    val outerOpacity = 0.18f - 0.13f * phase
    val outerWidth = 4.dp.toPx() + 4.dp.toPx() * phase
    drawCircle(
        color = BaoziTheme.accent.copy(alpha = outerOpacity),
        radius = radius * scale,
        center = targetRect.center,
        style = Stroke(width = outerWidth),
    )
    drawCircle(
        color = BaoziTheme.accent.copy(alpha = 0.7f),
        radius = radius * scale,
        center = targetRect.center,
        style = Stroke(width = 1.5.dp.toPx()),
    )
}

private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawCoachmarkArrow(
    from: Offset,
    to: Offset,
    targetRect: Rect,
    labelRect: Rect,
    style: CoachmarkLineStyle,
    isPrimary: Boolean,
) {
    val start = trimToRect(from = to, toward = from, rect = labelRect.inflate(6f))
    val endInset = if (style == CoachmarkLineStyle.SmoothCurve) -11f else -6f
    val end = trimToRect(from = from, toward = to, rect = targetRect.inflate(endInset))
    val dx = end.x - start.x
    val dy = end.y - start.y
    val length = max(1f, sqrt(dx * dx + dy * dy))
    val ux = dx / length
    val uy = dy / length
    val nx = -uy
    val ny = ux
    val path = when (style) {
        CoachmarkLineStyle.SmoothCurve -> makeSmoothCurve(start, end, nx, ny, length)
        CoachmarkLineStyle.SolidSquiggle,
        CoachmarkLineStyle.DashedSquiggle,
        CoachmarkLineStyle.Dotted,
        -> makeSquigglePath(
            start = start,
            end = end,
            dx = dx,
            dy = dy,
            nx = nx,
            ny = ny,
            length = length,
            amplitude = squiggleAmplitude(style, length),
        )
    }
    drawPath(
        path = path,
        color = BaoziTheme.accent.copy(alpha = if (isPrimary) 0.95f else 0.80f),
        style = strokeStyle(style, isPrimary),
    )

    val headLen = 8f
    val headHalfWidth = 4.5f
    val baseX = end.x - ux * headLen
    val baseY = end.y - uy * headLen
    val leftX = baseX + (-uy) * headHalfWidth
    val leftY = baseY + ux * headHalfWidth
    val rightX = baseX - (-uy) * headHalfWidth
    val rightY = baseY - ux * headHalfWidth
    val head = Path().apply {
        moveTo(end.x, end.y)
        lineTo(leftX, leftY)
        lineTo(rightX, rightY)
        close()
    }
    drawPath(head, BaoziTheme.accent.copy(alpha = if (isPrimary) 0.95f else 0.9f))
}

private fun squiggleAmplitude(style: CoachmarkLineStyle, length: Float): Float {
    if (length <= 70f) return 0f
    return when (style) {
        CoachmarkLineStyle.SolidSquiggle -> 5.5f
        CoachmarkLineStyle.DashedSquiggle -> 4.5f
        CoachmarkLineStyle.Dotted -> 0f
        CoachmarkLineStyle.SmoothCurve -> 0f
    }
}

private fun strokeStyle(style: CoachmarkLineStyle, isPrimary: Boolean): Stroke {
    return when (style) {
        CoachmarkLineStyle.SmoothCurve -> Stroke(width = 1.4f, cap = StrokeCap.Round, join = StrokeJoin.Round)
        CoachmarkLineStyle.SolidSquiggle -> Stroke(
            width = if (isPrimary) 1.8f else 1.4f,
            cap = StrokeCap.Round,
            join = StrokeJoin.Round,
        )
        CoachmarkLineStyle.DashedSquiggle -> Stroke(
            width = 1.4f,
            cap = StrokeCap.Round,
            join = StrokeJoin.Round,
            pathEffect = PathEffect.dashPathEffect(floatArrayOf(5f, 5f), 0f),
        )
        CoachmarkLineStyle.Dotted -> Stroke(
            width = 2.4f,
            cap = StrokeCap.Round,
            join = StrokeJoin.Round,
            pathEffect = PathEffect.dashPathEffect(floatArrayOf(0.01f, 7f), 0f),
        )
    }
}

private fun makeSquigglePath(
    start: Offset,
    end: Offset,
    dx: Float,
    dy: Float,
    nx: Float,
    ny: Float,
    length: Float,
    amplitude: Float,
): Path {
    val waves = max(2f, length / 44f)
    val steps = max(40, (length / 3f).roundToInt())
    return Path().apply {
        moveTo(start.x, start.y)
        for (i in 1..steps) {
            val t = i.toFloat() / steps.toFloat()
            val baseX = start.x + dx * t
            val baseY = start.y + dy * t
            val taper = sin((PI * t).toFloat())
            val phase = t * waves * 2f * PI.toFloat()
            val amp = amplitude * sin(phase) * taper
            lineTo(baseX + nx * amp, baseY + ny * amp)
        }
    }
}

private fun makeSmoothCurve(
    start: Offset,
    end: Offset,
    nx: Float,
    ny: Float,
    length: Float,
): Path {
    val mid = Offset((start.x + end.x) / 2f, (start.y + end.y) / 2f)
    val bend = min(7f, length * 0.06f)
    val control = Offset(mid.x + nx * bend, mid.y + ny * bend)
    return Path().apply {
        moveTo(start.x, start.y)
        quadraticTo(control.x, control.y, end.x, end.y)
    }
}

private fun trimToRect(from: Offset, toward: Offset, rect: Rect): Offset {
    val dx = from.x - toward.x
    val dy = from.y - toward.y
    val length = max(1f, sqrt(dx * dx + dy * dy))
    val ux = dx / length
    val uy = dy / length
    var t = 0f
    var probe = toward
    while (t <= length) {
        if (!rect.contains(probe)) return probe
        probe = Offset(probe.x + ux, probe.y + uy)
        t += 1f
    }
    return from
}

private fun Rect.inflate(amount: Float): Rect {
    return Rect(left - amount, top - amount, right + amount, bottom + amount)
}

File: apps/ios/Sources/Baozi/Views/OnboardingCoachmarks.swift
import SwiftUI

enum CoachmarkTarget: Hashable {
    case addServer
    case newThread
    case search
    case voice
}

struct CoachmarkAnchorKey: PreferenceKey {
    static var defaultValue: [CoachmarkTarget: Anchor<CGRect>] = [:]
    static func reduce(
        value: inout [CoachmarkTarget: Anchor<CGRect>],
        nextValue: () -> [CoachmarkTarget: Anchor<CGRect>]
    ) {
        value.merge(nextValue(), uniquingKeysWith: { _, new in new })
    }
}

extension View {
    func coachmarkAnchor(_ target: CoachmarkTarget) -> some View {
        anchorPreference(key: CoachmarkAnchorKey.self, value: .bounds) { anchor in
            [target: anchor]
        }
    }
}

struct OnboardingCoachmarksView: View {
    let anchors: [CoachmarkTarget: Anchor<CGRect>]

    @Environment(\.colorScheme) private var colorScheme

    enum LineStyle {
        case smoothCurve        // straight quadratic curve, no wiggle
        case solidSquiggle      // continuous sine wave
        case dashedSquiggle     // dashed sine wave
        case dotted             // round dot pattern
    }

    private struct Item: Identifiable {
        let id: CoachmarkTarget
        let primary: String
        let secondary: String?
        /// Label center as a fraction of the container size (0..1 in each
        /// axis). The bottom three labels sit close to their target buttons
        /// so the arrows are short; addServer sits below the pill with
        /// enough vertical gap that the arrow's curve is visible.
        let position: UnitPoint
        /// Fixed render width used both for layout and arrow trimming. Kept
        /// narrow so labels don't bleed into each other's vertical bands.
        let labelWidth: CGFloat
        let labelAlignment: HorizontalAlignment
        let style: LineStyle
        let isPrimary: Bool
    }

    /// Layout principle: each arrow lives in its own horizontal x-column
    /// (target's x), and each label sits in a vertical band where no other
    /// arrow passes through it. Bottom three labels are pulled down close to
    /// their target buttons so the arrows are short and readable.
    private let items: [Item] = [
        Item(
            id: .addServer,
            primary: "添加远程电脑",
            secondary: "有的话再加",
            position: UnitPoint(x: 0.55, y: 0.20),
            labelWidth: 200,
            labelAlignment: .center,
            style: .smoothCurve,
            isPrimary: false
        ),
        Item(
            id: .search,
            primary: "查看\n全部会话",
            secondary: nil,
            position: UnitPoint(x: 0.92, y: 0.62),
            labelWidth: 110,
            labelAlignment: .trailing,
            style: .dashedSquiggle,
            isPrimary: false
        ),
        Item(
            id: .newThread,
            primary: "新建会话",
            secondary: "或直接输入消息",
            position: UnitPoint(x: 0.50, y: 0.70),
            labelWidth: 220,
            labelAlignment: .center,
            style: .solidSquiggle,
            isPrimary: true
        ),
        Item(
            id: .voice,
            primary: "实时语音",
            secondary: "需在设置填 OpenAI 密钥",
            position: UnitPoint(x: 0.20, y: 0.78),
            labelWidth: 160,
            labelAlignment: .leading,
            style: .dotted,
            isPrimary: false
        )
    ]

    var body: some View {
        GeometryReader { proxy in
            ZStack(alignment: .topLeading) {
                ForEach(items) { item in
                    if let rect = resolvedRect(for: item.id, in: proxy) {
                        coachmark(for: item, targetRect: rect, container: proxy.size)
                    }
                }
            }
            .frame(width: proxy.size.width, height: proxy.size.height, alignment: .topLeading)
        }
        .allowsHitTesting(false)
    }

    /// Resolve the on-screen rect for a target. Server pill, +, and search
    /// publish anchors from inside the dashboard subtree. The voice button
    /// is overlaid externally (in `BaoziApp.primaryNavigationStack`) so its
    /// anchor never reaches us — fall back to a fixed position matching
    /// `homeVoiceLauncher`'s `.bottomLeading` placement (14pt leading inset,
    /// 4pt bottom inset, 44pt button size).
    private func resolvedRect(for target: CoachmarkTarget, in proxy: GeometryProxy) -> CGRect? {
        if let anchor = anchors[target] {
            return proxy[anchor]
        }
        if target == .voice {
            let size: CGFloat = 44
            let leading: CGFloat = 14
            let bottomInset: CGFloat = 4
            return CGRect(
                x: leading,
                y: proxy.size.height - bottomInset - size,
                width: size,
                height: size
            )
        }
        return nil
    }

    @ViewBuilder
    private func coachmark(for item: Item, targetRect: CGRect, container: CGSize) -> some View {
        let labelHeight: CGFloat = 60
        let target = CGPoint(x: targetRect.midX, y: targetRect.midY)

        // Place the label at its assigned absolute position, then clamp to
        // the container so it never overflows on a narrow device.
        let proposedX = container.width * item.position.x - item.labelWidth / 2
        let proposedY = container.height * item.position.y - labelHeight / 2
        let clampedX = min(max(8, proposedX), container.width - item.labelWidth - 8)
        let clampedY = min(max(8, proposedY), container.height - labelHeight - 8)
        let labelRect = CGRect(x: clampedX, y: clampedY, width: item.labelWidth, height: labelHeight)
        let labelCenter = CGPoint(x: labelRect.midX, y: labelRect.midY)

        ZStack(alignment: .topLeading) {
            // Halo ring for the primary target.
            if item.isPrimary {
                CoachmarkHalo()
                    .frame(width: targetRect.width + 18, height: targetRect.height + 18)
                    .position(x: target.x, y: target.y)
                    // Disable implicit animation on the position itself — the
                    // halo's TimelineView pulse must NOT trickle into the
                    // .position update, otherwise tiny anchor jitter from the
                    // glass-morph button gets interpolated across the full
                    // animation curve and reads as the halo flying around.
                    .transaction { $0.animation = nil }
            }

            CoachmarkArrow(
                from: labelCenter,
                to: target,
                targetRect: targetRect,
                labelRect: labelRect,
                style: item.style,
                isPrimary: item.isPrimary
            )

            VStack(alignment: item.labelAlignment, spacing: 2) {
                Text(item.primary)
                    .baoziMonoFont(size: 12, weight: .semibold)
                    .foregroundStyle(BaoziTheme.accent)
                    .multilineTextAlignment(textAlignment(for: item.labelAlignment))
                    .fixedSize(horizontal: false, vertical: true)
                if let secondary = item.secondary {
                    Text(secondary)
                        .baoziMonoFont(size: 10, weight: .regular)
                        .foregroundStyle(BaoziTheme.textSecondary)
                        .multilineTextAlignment(textAlignment(for: item.labelAlignment))
                        .fixedSize(horizontal: false, vertical: true)
                }
            }
            .shadow(color: coachmarkTextShadowColor, radius: coachmarkTextShadowRadius, x: 0, y: 1)
            .frame(width: item.labelWidth, height: labelHeight, alignment: vstackFrameAlignment(for: item.labelAlignment))
            .position(x: labelCenter.x, y: labelCenter.y)
        }
    }

    private var coachmarkTextShadowColor: Color {
        colorScheme == .dark ? .black.opacity(0.7) : .clear
    }

    private var coachmarkTextShadowRadius: CGFloat {
        colorScheme == .dark ? 4 : 0
    }

    private func textAlignment(for h: HorizontalAlignment) -> TextAlignment {
        switch h {
        case .leading:  return .leading
        case .trailing: return .trailing
        default:        return .center
        }
    }

    private func vstackFrameAlignment(for h: HorizontalAlignment) -> Alignment {
        switch h {
        case .leading:  return .topLeading
        case .trailing: return .topTrailing
        default:        return .top
        }
    }
}

/// Pulsing ring around the primary target. Uses `TimelineView` to drive the
/// animation from a wall-clock so SwiftUI's animation system never gets
/// involved — earlier `.animation(.repeatForever, value: pulse)` left an
/// active animation context in this subtree forever, causing every other
/// property change in the parent (notably `.position`) to interpolate over
/// the same 1.2s curve and visually fly across the screen.
private struct CoachmarkHalo: View {
    var body: some View {
        TimelineView(.animation(minimumInterval: 1.0 / 30.0)) { context in
            // Phase: 0..1, period 1.6s, sinusoidal so it eases at endpoints.
            let t = context.date.timeIntervalSinceReferenceDate
            let phase = (sin(t * 2 * .pi / 1.6) + 1) / 2

            let scale = 1.0 + 0.08 * phase
            let outerOpacity = 0.18 - 0.13 * phase
            let outerWidth = 4 + 4 * phase

            Circle()
                .stroke(BaoziTheme.accent.opacity(0.7), lineWidth: 1.5)
                .background(
                    Circle()
                        .stroke(BaoziTheme.accent.opacity(outerOpacity), lineWidth: outerWidth)
                        .blur(radius: 2)
                )
                .scaleEffect(scale)
        }
    }
}

/// Hand-drawn-feeling coachmark arrow. The path is either a smooth quadratic
/// curve (`smoothCurve`) or a sine-wave squiggle along the straight line
/// between `from` and `to`. Stroke style varies per item: solid, dashed, or
/// dotted (round-cap zero-length dashes). The wiggle amplitude tapers to
/// zero at each end so the arrowhead direction stays clean.
private struct CoachmarkArrow: View {
    let from: CGPoint
    let to: CGPoint
    let targetRect: CGRect
    let labelRect: CGRect
    let style: OnboardingCoachmarksView.LineStyle
    let isPrimary: Bool

    var body: some View {
        Canvas { ctx, _ in
            // Trim so the path starts outside the label box and ends with
            // clearance from the button so the arrowhead doesn't poke into
            // it. The target gap is wider for short arrows because the
            // arrowhead occupies a larger fraction of the visible line.
            let start = trimToRect(from: to, toward: from, rect: labelRect.insetBy(dx: -6, dy: -6))
            let endInset: CGFloat = (style == .smoothCurve) ? -11 : -6
            let end = trimToRect(from: from, toward: to, rect: targetRect.insetBy(dx: endInset, dy: endInset))

            let dx = end.x - start.x
            let dy = end.y - start.y
            let length = max(1, sqrt(dx * dx + dy * dy))
            let ux = dx / length
            let uy = dy / length
            // Perpendicular unit vector (rotate 90°). Stable sign keeps the
            // wiggle on a consistent side of the line.
            let nx = -uy
            let ny = ux

            let path: Path
            switch style {
            case .smoothCurve:
                path = makeSmoothCurve(start: start, end: end, nx: nx, ny: ny, length: length)
            case .solidSquiggle, .dashedSquiggle, .dotted:
                let amplitude: CGFloat = squiggleAmplitude(for: style, length: length)
                path = makeSquigglePath(
                    start: start,
                    end: end,
                    dx: dx, dy: dy,
                    nx: nx, ny: ny,
                    length: length,
                    amplitude: amplitude
                )
            }

            ctx.stroke(
                path,
                with: .color(BaoziTheme.accent.opacity(isPrimary ? 0.95 : 0.80)),
                style: strokeStyle(for: style, isPrimary: isPrimary)
            )

            // Arrowhead aligned with the underlying straight direction (the
            // squiggle's amplitude is zero at the end thanks to the taper,
            // so this matches the painted line's true tangent at the tip).
            let headLen: CGFloat = 8
            let headHalfWidth: CGFloat = 4.5
            let baseX = end.x - ux * headLen
            let baseY = end.y - uy * headLen
            let leftX = baseX + (-uy) * headHalfWidth
            let leftY = baseY + ux * headHalfWidth
            let rightX = baseX - (-uy) * headHalfWidth
            let rightY = baseY - ux * headHalfWidth

            var head = Path()
            head.move(to: end)
            head.addLine(to: CGPoint(x: leftX, y: leftY))
            head.addLine(to: CGPoint(x: rightX, y: rightY))
            head.closeSubpath()
            ctx.fill(head, with: .color(BaoziTheme.accent.opacity(isPrimary ? 0.95 : 0.9)))
        }
    }

    private func squiggleAmplitude(for style: OnboardingCoachmarksView.LineStyle, length: CGFloat) -> CGFloat {
        // Skip squiggle on short arrows — looks like noise at small scale.
        guard length > 70 else { return 0 }
        switch style {
        case .solidSquiggle:  return 5.5
        case .dashedSquiggle: return 4.5
        case .dotted:         return 0       // straight dotted line — no waves
        case .smoothCurve:    return 0
        }
    }

    private func strokeStyle(for style: OnboardingCoachmarksView.LineStyle, isPrimary: Bool) -> StrokeStyle {
        switch style {
        case .smoothCurve:
            return StrokeStyle(lineWidth: 1.4, lineCap: .round, lineJoin: .round)
        case .solidSquiggle:
            return StrokeStyle(lineWidth: isPrimary ? 1.8 : 1.4, lineCap: .round, lineJoin: .round)
        case .dashedSquiggle:
            return StrokeStyle(lineWidth: 1.4, lineCap: .round, lineJoin: .round, dash: [5, 5])
        case .dotted:
            // dash[0] near 0 + round cap → renders as round dots spaced by dash[1].
            return StrokeStyle(lineWidth: 2.4, lineCap: .round, lineJoin: .round, dash: [0.01, 7])
        }
    }

    private func makeSquigglePath(
        start: CGPoint, end: CGPoint,
        dx: CGFloat, dy: CGFloat,
        nx: CGFloat, ny: CGFloat,
        length: CGFloat, amplitude: CGFloat
    ) -> Path {
        let waves: CGFloat = max(2, length / 44)
        let steps = max(40, Int(length / 3))
        var path = Path()
        path.move(to: start)
        for i in 1...steps {
            let t = CGFloat(i) / CGFloat(steps)
            let baseX = start.x + dx * t
            let baseY = start.y + dy * t
            // Hann-window taper: 0 at t=0 and t=1, peak at t=0.5.
            let taper = sin(.pi * t)
            let phase = t * waves * 2 * .pi
            let amp = amplitude * sin(phase) * taper
            path.addLine(to: CGPoint(x: baseX + nx * amp, y: baseY + ny * amp))
        }
        return path
    }

    private func makeSmoothCurve(
        start: CGPoint, end: CGPoint,
        nx: CGFloat, ny: CGFloat,
        length: CGFloat
    ) -> Path {
        // Subtle curve only. The earlier `length * 0.18` produced a wide arc
        // that made short arrows (like addServer's ~80pt run to the pill)
        // sweep way out to the side before terminating.
        let mid = CGPoint(x: (start.x + end.x) / 2, y: (start.y + end.y) / 2)
        let bend = min(7, length * 0.06)
        let control = CGPoint(x: mid.x + nx * bend, y: mid.y + ny * bend)
        var path = Path()
        path.move(to: start)
        path.addQuadCurve(to: end, control: control)
        return path
    }

    /// Walks from `from` toward `target` and returns the first point that
    /// falls outside `rect`'s perimeter — used to clip the line so it
    /// neither enters the label box nor overshoots into the target.
    private func trimToRect(from target: CGPoint, toward from: CGPoint, rect: CGRect) -> CGPoint {
        let dx = target.x - from.x
        let dy = target.y - from.y
        let length = max(1, sqrt(dx * dx + dy * dy))
        let ux = dx / length
        let uy = dy / length

        var t: CGFloat = 0
        let step: CGFloat = 1
        var probe = from
        let maxT = length
        while t <= maxT {
            if !rect.contains(probe) { return probe }
            probe.x += ux * step
            probe.y += uy * step
            t += step
        }
        return target
    }
}
